<a href="https://colab.research.google.com/github/awfajri/artificial-intelligence/blob/main/Tugas_Praktikum_6_Auf_Fajri_Ramadhani.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Tugas Praktikum Bab 6
<br>
Nama : Auf Fajri Ramadhani
<br>
NPM  : 2410631170059

#Tugas Jaringan Syaraf Tiruan - Modul Bab 6
Soal:
<br>
Latihan: Uji Coba Menggunakan Berbagai Jenis Fungsi Aktivasi (Membership Function)

1.   Gunakan fungsi aktivasi linear dan hyperbolic tangent untuk data yang sama dan tunjukkan hasilnya!
2.   Apakah terjadi perbedaan hasil pelatihan dan prediksi pada saat menggunakan fungsi aktivasi yang berbeda? Tunjukkan dan jelaskan! (dapat menggunakan tabel untuk menunjukkan nilai perbedaannya)
3. Gunakan variasi epoch (2500, 3500, 4500) dan tunjukkan hasil pelatihan dan hasil prediksinya! (dapat menggunakan tabel untuk menunjukkan nilai perbedaannya)



#Jawab

#Langkah 1 Import library dan dataset

In [ ]:
import numpy as np

input_set = np.array([[0,1,0],
                      [0,0,1],
                      [1,0,0],
                      [1,1,0],
                      [1,1,1],
                      [0,1,1],
                      [0,1,0]])

labels = np.array([[1, 0, 0, 1, 1, 0, 1]]).reshape(7,1)

def reset_weights():
    np.random.seed(42)
    weights = np.random.rand(3,1)
    bias = np.random.rand(1)
    return weights, bias

lr = 0.05

Langkah pertama, saya mengimpor library numpy untuk membantu perhitungan komputasi matriks. Kemudian, saya menyiapkan input_set dan labels sesuai dengan modul. Saya juga membuat sebuah fungsi bernama reset_weights() agar setiap kali kita menguji model dengan fungsi aktivasi atau epoch yang baru, bobot awalnya selalu sama dan perbandingannya menjadi adil.

#Langkah 2: Fungsi Aktivasi Linear dan Hyperbolic Tangent (soal 1)

In [ ]:
#1A. Fungsi aktivasi linear
def linear(x):
    return x

def linear_derivative(x):
    return 1

weights, bias = reset_weights()

print("1. train dengan fungsi linear")
for epoch in range(25000):
    inputs = input_set
    XW = np.dot(inputs, weights) + bias
    z = linear(XW)
    error = z - labels

    dcost = error
    dpred = linear_derivative(z)
    z_del = dcost * dpred

    weights = weights - lr * np.dot(inputs.T, z_del)
    for num in z_del:
        bias = bias - lr * num

print("Prediksi Linear [1,0,0]:", linear(np.dot(np.array([1,0,0]), weights) + bias)[0])
print("Prediksi Linear [0,1,0]:", linear(np.dot(np.array([0,1,0]), weights) + bias)[0])


#1B. Fungsi Aktivasi hyperbolic tangent (Tanh)
def tanh(x):
    return np.tanh(x)

def tanh_derivative(x):
    return 1 - np.tanh(x)**2

weights, bias = reset_weights()

print("\n2. train dengan fungsi Tanh")
for epoch in range(25000):
    inputs = input_set
    XW = np.dot(inputs, weights) + bias
    z = tanh(XW)
    error = z - labels

    dcost = error
    dpred = tanh_derivative(XW)
    z_del = dcost * dpred

    weights = weights - lr * np.dot(inputs.T, z_del)
    for num in z_del:
        bias = bias - lr * num

print("Prediksi Tanh [1,0,0]:", tanh(np.dot(np.array([1,0,0]), weights) + bias)[0])
print("Prediksi Tanh [0,1,0]:", tanh(np.dot(np.array([0,1,0]), weights) + bias)[0])

#1C. Fungsi sigmoid
def sigmoid(x):
  return 1/(1+np.exp(-x))

def sigmoid_derivative(x):
  return sigmoid (x)*(1-sigmoid(x))

weights, bias = reset_weights()

print("\n3. train dengan fungsi sigmoid")
for epoch in range(25000):
  inputs = input_set
  XW = np.dot(inputs, weights) + bias
  z = sigmoid(XW)
  error = z - labels

  dcost = error
  dpred = sigmoid_derivative(XW)
  z_del = dcost * dpred

  weights = weights - lr * np.dot(inputs.T, z_del)
  for num in z_del:
      bias = bias - lr * num

print("Prediksi sigmoid [1,0,0]:", sigmoid(np.dot(np.array([1,0,0]), weights) + bias)[0])
print("Prediksi sigmoid [0,1,0]:", sigmoid(np.dot(np.array([0,1,0]), weights) + bias)[0])


1. train dengan fungsi linear
Prediksi Linear [1,0,0]: 0.24999999999999933
Prediksi Linear [0,1,0]: 0.857142857142857

2. train dengan fungsi Tanh
Prediksi Tanh [1,0,0]: 0.23002210252305308
Prediksi Tanh [0,1,0]: 0.7853550628309144

3. train dengan fungsi sigmoid
Prediksi sigmoid [1,0,0]: 0.059303513977400565
Prediksi sigmoid [0,1,0]: 0.961753966725296


Selanjutnya untuk menjawab soal nomor 1, saya membuat blok kode untuk masing-masing fungsi aktivasi yaitu Linear, Tanh, dan Sigmoid. Di setiap bloknya, saya mendefinisikan rumus utama beserta fungsi turunannya karena jaringan saraf tiruan sangat membutuhkan nilai turunan ini untuk mengetahui seberapa besar kesalahan yang harus dikoreksi. Saya juga memanggil fungsi reset_weights() sebelum proses pelatihan dimulai agar nilai bobot dan bias selalu kembali diacak dari awal. Hal ini saya lakukan supaya ketiga fungsi tersebut bisa diuji secara adil dari titik start yang sama.

Selanjutnya, saya membuat perulangan sebanyak 25.000 kali (epoch) untuk melatih modelnya. Di dalam proses pelatihan ini, mesin akan mengalikan matriks input dengan bobot, lalu memprosesnya melalui fungsi aktivasi untuk menghasilkan sebuah tebakan. Setelah itu, mesin akan menghitung seberapa jauh tebakannya dari label yang sebenarnya untuk mencari tahu nilai error. Nilai error inilah yang kemudian dikalikan dengan turunan fungsi aktivasinya untuk memperbarui bobot mesin secara perlahan sesuai dengan nilai learning rate.

Langkah terakhir, setelah mesin selesai berlatih puluhan ribu kali, saya mengujinya dengan memberikan data baru yaitu array [1,0,0] dan [0,1,0]. Hasil tebakan akhir dari masing-masing fungsi aktivasi langsung saya tampilkan menggunakan perintah print. Dari output ini, kita nantinya bisa membandingkan secara langsung fungsi mana yang belajarnya paling maksimal dan tebakannya paling mendekati nilai target 0 dan 1.

#Analisis Perbedaan Fungsi Aktivasi (Soal 2)

betul, terjadi perbedaan hasil pelatihan dan prediksi yang sangat jelas pada saat menggunakan fungsi aktivasi yang berbeda. Berdasarkan hasil training dengan 25.000 epoch, berikut adalah tabel perbandingannya:

In [ ]:
import pandas as pd

data_hasil = {
    'Fungsi Aktivasi': ['Linear', 'Tanh', 'Sigmoid'],
    'Prediksi [1,0,0] (Target: 0)': [0.2499, 0.2300, 0.0593],
    'Prediksi [0,1,0] (Target: 1)': [0.8571, 0.7853, 0.9617]
}

df_hasil = pd.DataFrame(data_hasil)

df_hasil

,Fungsi Aktivasi,"Prediksi [1,0,0] (Target: 0)","Prediksi [0,1,0] (Target: 1)"
0,Linear,0.2499,0.8571
1,Tanh,0.2300,0.7853
2,Sigmoid,0.0593,0.9617


Kesimpulan untuk Soal 2:
<br>
Perbedaan nilai di atas terjadi karena sifat matematis dari masing-masing fungsi aktivasi:
<br>


1.   Fungsi Linear hanya meneruskan nilai penjumlahan bobot tanpa membatasinya (berupa garis lurus). Akibatnya, model kesulitan untuk membedakan data klasifikasi 0 dan 1 secara tajam, sehingga menghasilkan nilai tebakan yang "mengambang" (seperti 0.24 dan 0.85).
2.   Fungsi Tanh memiliki rentang nilai dari -1 hingga 1. Meskipun lebih baik dalam menangani data non-linear, rentang nilainya membuat konvergensi ke angka 0 menjadi sedikit lambat untuk kasus klasifikasi biner ini.
3. Fungsi Sigmoid adalah yang terbaik untuk kasus ini karena ia dirancang khusus untuk memadatkan (squash) nilai hasil perhitungan ke dalam rentang probabilitas yang sangat ketat antara 0 hingga 1. Hal ini terlihat dari hasil prediksinya (0.059 dan 0.961) yang sangat meyakinkan dan paling mendekati label aslinya.



#3. Variasi Epoch 2500, 3500, 4500 (soal 3)

Disini saya mendefinisikan dulu nilai epoch yg akan digunakan nanti untuk pengujian pada masing-masing fungsi.

In [ ]:
epochs_to_test = [2500, 3500, 4500]

#Pengujian Fungsi Linear

In [ ]:
def linear(x): return x
def linear_derivative(x): return 1

print("=== PENGUJIAN VARIASI EPOCH: LINEAR ===")
for ep in epochs_to_test:
    weights, bias = reset_weights()
    for _ in range(ep):
        z = linear(np.dot(input_set, weights) + bias)
        z_del = (z - labels) * linear_derivative(z)
        weights -= lr * np.dot(input_set.T, z_del)
        for num in z_del: bias -= lr * num

    p1 = linear(np.dot([1,0,0], weights) + bias)[0]
    p2 = linear(np.dot([0,1,0], weights) + bias)[0]
    print(f"[Epoch {ep}] Error: {abs((z-labels).sum()):.4f} | Prediksi [1,0,0]: {p1:.4f} | Prediksi [0,1,0]: {p2:.4f}")

=== PENGUJIAN VARIASI EPOCH: LINEAR ===
[Epoch 2500] Error: 0.0000 | Prediksi [1,0,0]: 0.2500 | Prediksi [0,1,0]: 0.8571
[Epoch 3500] Error: 0.0000 | Prediksi [1,0,0]: 0.2500 | Prediksi [0,1,0]: 0.8571
[Epoch 4500] Error: 0.0000 | Prediksi [1,0,0]: 0.2500 | Prediksi [0,1,0]: 0.8571


Pada pengujian pertama menggunakan fungsi Linear, saya melihat hasil yang cukup unik. Nilai pergerakan error menunjukkan angka 0.0000 dan hasil tebakan untuk ketiga variasi epoch (2500, 3500, 4500) sama sekali tidak berubah, yaitu tertahan di angka 0.2500 dan 0.8571. Ini menunjukkan bahwa model Jaringan Saraf Tiruan sudah mencapai titik mentoknya (konvergen) jauh sebelum menyentuh 2500 epoch. Penambahan waktu belajar hingga 4500 iterasi pun tidak lagi memberikan efek apa-apa karena fungsi linear pada dasarnya hanya menarik garis komputasi lurus, sehingga ia tidak memiliki fleksibilitas untuk membedakan data target biner (0 dan 1) secara mutlak.

#Pengujian Fungsi Tanh

In [ ]:
def tanh(x): return np.tanh(x)
def tanh_derivative(x): return 1 - np.tanh(x)**2

print("=== PENGUJIAN VARIASI EPOCH: TANH ===")
for ep in epochs_to_test:
    weights, bias = reset_weights()
    for _ in range(ep):
        XW = np.dot(input_set, weights) + bias
        z = tanh(XW)
        z_del = (z - labels) * tanh_derivative(XW)
        weights -= lr * np.dot(input_set.T, z_del)
        for num in z_del: bias -= lr * num

    p1 = tanh(np.dot([1,0,0], weights) + bias)[0]
    p2 = tanh(np.dot([0,1,0], weights) + bias)[0]
    print(f"[Epoch {ep}] Error: {abs((z-labels).sum()):.4f} | Prediksi [1,0,0]: {p1:.4f} | Prediksi [0,1,0]: {p2:.4f}")

=== PENGUJIAN VARIASI EPOCH: TANH ===
[Epoch 2500] Error: 0.2970 | Prediksi [1,0,0]: 0.2300 | Prediksi [0,1,0]: 0.7854
[Epoch 3500] Error: 0.2970 | Prediksi [1,0,0]: 0.2300 | Prediksi [0,1,0]: 0.7854
[Epoch 4500] Error: 0.2970 | Prediksi [1,0,0]: 0.2300 | Prediksi [0,1,0]: 0.7854


Kejadian serupa juga saya temukan pada pengujian fungsi aktivasi Hyperbolic Tangent (Tanh). Meskipun iterasi belajarnya sudah saya naikkan secara bertahap hingga 4500 epoch, nilai sisa error tetap mandek di angka 0.2970. Begitu pula dengan nilai prediksinya yang sama sekali tidak bergeser dari angka 0.2300 untuk input pertama dan 0.7854 untuk input kedua. Dalam kasus arsitektur jaringan yang kita bangun ini, pembaharuan bobot pada fungsi Tanh kemungkinan sudah berhenti berkembang (gradient vanishing) di awal iterasi. Akibatnya, menambah jumlah epoch terbukti tidak lagi berdampak pada peningkatan akurasi tebakan mesin.

#Pengujian Fungsi Sigmoid

In [ ]:
def sigmoid(x): return 1/(1+np.exp(-x))
def sigmoid_derivative(x):
  return sigmoid (x)*(1-sigmoid(x))

print("=== PENGUJIAN VARIASI EPOCH: SIGMOID ===")
for ep in epochs_to_test:
    weights, bias = reset_weights()
    for _ in range(ep):
        z = sigmoid(np.dot(input_set, weights) + bias)
        z_del = (z - labels) * sigmoid_derivative(z)
        weights -= lr * np.dot(input_set.T, z_del)
        for num in z_del: bias -= lr * num

    p1 = sigmoid(np.dot([1,0,0], weights) + bias)[0]
    p2 = sigmoid(np.dot([0,1,0], weights) + bias)[0]
    print(f"[Epoch {ep}] Error: {abs((z-labels).sum()):.4f} | Prediksi [1,0,0]: {p1:.4f} | Prediksi [0,1,0]: {p2:.4f}")

=== PENGUJIAN VARIASI EPOCH: SIGMOID ===
[Epoch 2500] Error: 0.0167 | Prediksi [1,0,0]: 0.1667 | Prediksi [0,1,0]: 0.9212
[Epoch 3500] Error: 0.0067 | Prediksi [1,0,0]: 0.1319 | Prediksi [0,1,0]: 0.9350
[Epoch 4500] Error: 0.0018 | Prediksi [1,0,0]: 0.1089 | Prediksi [0,1,0]: 0.9449


Hasil yang paling ideal dan dinamis akhirnya saya dapatkan pada pengujian fungsi Sigmoid. Berbeda dengan dua fungsi sebelumnya yang jalan di tempat, penambahan jumlah epoch di sini benar-benar memperlihatkan proses belajar mesin yang nyata. Saat saya menaikkan epochnya dari 2500 berturut-turut hingga 4500, sisa error-nya terus menyusut secara konsisten dari 0.0167 turun menjadi 0.0018. Hasil prediksinya pun semakin tajam; tebakan untuk target 0 terus menurun perlahan ke angka 0.1089, dan tebakan untuk target 1 semakin naik merapat ke angka 0.9449. Eksperimen ini membuktikan bahwa fungsi Sigmoid sangat cocok untuk kasus klasifikasi biner, di mana memperbanyak epoch sangat terbukti efektif menajamkan akurasi kecerdasan buatan.